### PR TITLE
parameterize some tests

### PR DESCRIPTION
**Pull request by ferrants**
Thank you for sending your pull request. But first, have you included
unit tests, and is your code PEP8 conformant? [More details](https://github.com/freqtrade/freqtrade/blob/develop/CONTRIBUTING.md)

## Summary

Parameterize some tests, per @samgermain 's request

Solve the issue: None

## Quick changelog

- Parameterized some tests

## What's new?

Fewer tests, reduced code, easier to add new tests

### CODE REVIEW

File: tests/test_freqtradebot.py
Comment (ferrants) [inline tests/test_freqtradebot.py]: The log assertions seem like they could be calculated, would that be preferred over asserting a static value?

File: tests/test_freqtradebot.py
Comment (ferrants) [inline tests/test_freqtradebot.py:3664]: It was unclear to me how the `0.0005` was determined here and why it didn't match the `expected_fee` or the `fee_cost`. So I left this as a parameter.

Comment (xmatthias) [reply tests/test_freqtradebot.py:3664]: You'll have to look at `trades_for_order3` (probably in `conftest.py`) for this.
This test is about gathering fees from a trade-result (because `fetch_order()` didn't contain fees) - and there were 2 trade necessary to fill your order.

I'm not 100% sure if combining all of them into one makes sense (must look at the details of how it's done), as the mocks will need to be different (return something different).

Comment (ferrants) [reply tests/test_freqtradebot.py:3664]: I actually meant to remove that fixture, it is only used in this test case, is built on top of the `trades_for_order2` and the fixture only changes the `trades_for_order2` so slightly that it's probably worth just setting as a parameter. Unless we prefer to keep these in fixtures, in which case parameterizing probably doesn't make sense.

https://github.com/freqtrade/freqtrade/blob/6fc770d97dd8d4380be3fb94e8f3dcedfa981cd7/tests/conftest.py#L1688-L1693

Comment (ferrants) [reply tests/test_freqtradebot.py:3664]: Removed that fixture

**Comment by samgermain** [comment]
@xmatthias This PR might be good actually, there's some more code that could be reduced, but maybe we can do that as we see it while converting the prices to USDT (which I've started in another branch)

@ferrants good job

File: tests/test_freqtradebot.py
Comment (xmatthias) [inline tests/test_freqtradebot.py]: What's the point of stake_currency here (as parameter, i understand it might need to be set as 

BTC seems wrong (considering trades are for LTC/ETH) - and changing it to ETH doesn't seem to have any effect (i also struggle to see where we set stake_currency to BTC in the old tests ...)

File: tests/test_freqtradebot.py
Comment (samgermain) [inline tests/test_freqtradebot.py]: It looks like the 2 strings in these test are identical, so just make the number at the end a parameter

File: tests/test_freqtradebot.py
Comment (ferrants) [reply tests/test_freqtradebot.py]: That's what is in the `default_conf` and since the old `test_get_real_amount_multi2` test was setting it to ETH, being more explicit in the consolidation seemed reasonable.

Comment (samgermain) [reply tests/test_freqtradebot.py]: > That's what is in the `default_conf` and since the old `test_get_real_amount_multi2` test was setting it to ETH, being more explicit in the consolidation seemed reasonable.

I would remove it

File: tests/test_freqtradebot.py
Comment (samgermain) [inline tests/test_freqtradebot.py:3562]: can you remove the text here also

File: tests/test_freqtradebot.py
Comment (samgermain) [inline tests/test_freqtradebot.py]: I would remove the `stake_currency` parameter all together, if it's no adding anything to the test

**Comment by samgermain** [comment]
@ferrants 

Do you know about running these tests from the command line?

```
pytest
flake8 freqtrade tests scripts
mypy freqtrade
isort
```

File: tests/test_freqtradebot.py
Comment (ferrants) [reply tests/test_freqtradebot.py]: If I remove it, the second one fails. The prior `test_get_real_amount_multi2` test expects the stake currency in the config to be ETH. I can remove it and change the expected `fee_reduction_amount`, but then it won't match the old `test_get_real_amount_multi2`

**Comment by ferrants** [comment]
@samgermain , yeah, I had run those locally when I originally submitted the PR

Do you guys use anything to automatically format? I've used `black` (https://pypi.org/project/black/) in many projects and it's pretty awesome.

File: tests/test_freqtradebot.py
Comment (samgermain) [reply tests/test_freqtradebot.py]: Can it be ETH for both?

**Comment by samgermain** [comment]
> Do you guys use anything to automatically format? I've used black (https://pypi.org/project/black/) in many projects and it's pretty awesome.

I use autopep8, i think there's plans for black in the future

**Comment by xmatthias** [comment]
> Do you guys use anything to automatically format? I've used `black` (https://pypi.org/project/black/) in many projects and it's pretty awesome.

Black is nice if you start a project with it - it's ... a pretty large change to run on a project this size at this point - and i'm not really comfortable with doing that at the moment (we have 2 pretty big changes in seperate branches - both of which would have the risk of 100000 merge-conflicts in case of applying black now).

We'll probably look at this again - but surely not before feat/short is merged - and probably not until my "async-tests" stuff is either progressed or thrown away ... :laughing:

File: tests/test_freqtradebot.py
Comment (ferrants) [reply tests/test_freqtradebot.py]: Yup, that works. Updated

File: tests/test_freqtradebot.py
Comment (ferrants) [reply tests/test_freqtradebot.py:3562]: Should I not assert it at all? Most of the previous test cases weren't asserting a log line. And the log line on line 3554 is in a different format, the `Fee for Trade` one. Simply passing in a value won't work for that case and then this would be the only version that even asserts. Does that make sense?

Comment (xmatthias) [reply tests/test_freqtradebot.py:3562]: I'd leave it as is to be honest.

piecing the logstring together might work - but will make changes to it VERY difficult for future editors.

Comment (xmatthias) [reply tests/test_freqtradebot.py:3562]: One additional request though - could you add comments to each parameter-line documenting what we're trying to test? 
Currently it's pretty difficult to determine what exactly is trying to be tested...

File: tests/test_freqtradebot.py
Comment (ferrants) [inline tests/test_freqtradebot.py:3771]: This case appears to be similar to `test_get_real_amount_no_currency_in_fee`, but checking if `currency` was not in the dict at all instead of set to None. This probably caught a regression or specific issue. Should we have the whole `fee` dict as a parameter instead of separate `fee_currency` and `fee_cost`, so we can enable this testing in a parameterized way?

File: tests/test_freqtradebot.py
Comment (ferrants) [inline tests/test_freqtradebot.py]: Now, this is essentially duplicate of https://github.com/freqtrade/freqtrade/pull/5601/files#diff-f0d1952492d82eb12612502e7426fa75194b47a35bf0a04a9ad1af0f924ee25bR3551

Mentioned more here:
https://github.com/freqtrade/freqtrade/pull/5601/files#r714345676

File: tests/test_freqtradebot.py
Comment (samgermain) [reply tests/test_freqtradebot.py:3771]: > This case appears to be similar to `test_get_real_amount_no_currency_in_fee`, but checking if `currency` was not in the dict at all instead of set to None. This probably caught a regression or specific issue. Should we have the whole `fee` dict as a parameter instead of separate `fee_currency` and `fee_cost`, so we can enable this testing in a parameterized way?

I kinda don't want to focus on parametrizing anymore, and get this merged, so that we can start focussing on USDT prices

Comment (xmatthias) [reply tests/test_freqtradebot.py:3771]: most of these tests will be quite similar - but all capture another edge-case we've encountered (exchange responses are not always the same - and there's a LOT of edge-cases you can encounter, especially when multiple trades filling one order are involved).

File: tests/test_freqtradebot.py
Comment (samgermain) [reply tests/test_freqtradebot.py]: i don't understand what your trying to say here

Comment (ferrants) [reply tests/test_freqtradebot.py]: This test case is the same as ` (0.004, None, 0, True, None),` This is no longer testing what it was originally testing. An option to keep it the same would be to pass in the fee object instead, `{"cost": 0.008}` and `{"cost": 0.004, "currency": None}` for the previous other test case to capture testing the same thing as before

Comment (xmatthias) [reply tests/test_freqtradebot.py]: I THINK that's the better way to test this, yes - it's also clearer to understand - IMHO

Comment (samgermain) [reply tests/test_freqtradebot.py]: > This test case is the same as ` (0.004, None, 0, True, None),` This is no longer testing what it was originally testing. An option to keep it the same would be to pass in the fee object instead, `{"cost": 0.008}` and `{"cost": 0.004, "currency": None}` for the previous other test case to capture testing the same thing as before

Sounds good, make it test what it was originally testing, I think that this can get merged after that

File: tests/test_freqtradebot.py
Comment (xmatthias) [inline tests/test_freqtradebot.py:3551]: While i updated these tests to use a "fee" dict now ... i've discovered that this testcase does not test the same thing than the existing test.

`test_get_real_amount_stake` (which updates currency to 'ETH') keeps the cost of 0.008 from the original trades_for_order - so a correct test (correctly replicating previous behaviour) would be `{'cost': 0.008, 'currency': 'ETH'}`.

I've now adjusted this test accordingly (check the last commit for necessary changes) ... and based on my comparisons, the remaining tests should be ok, so for me, this would be OK now.

Comment (samgermain) [reply tests/test_freqtradebot.py:3551]: ok, can you merge this if there's no other problems?

Comment (xmatthias) [reply tests/test_freqtradebot.py:3551]: i can - but wanted to give both you and the original author a chance to look at it.

Also - please stop pushing for merges or reviews. 
I know it takes patience to wait for reviews or merges, but to be honest, freqtrade is quite good in that regards (in other sometimes very active repositories (multiple merges per day) i've contributed to, you can wait for months for even an initial review).
It's putting unnecessary pressure on other people - which might be fine in a company where it's people's main work to do this - but it's not ok for open source projects, where it's people's free time that is invested.
Thanks for your understanding.

Comment (samgermain) [reply tests/test_freqtradebot.py:3551]: if it's fine by you it's fine by me

Comment (xmatthias) [reply tests/test_freqtradebot.py:3551]: Yes - but when editing some parts of user PR's - it's either small adjustments (Right before merging) - or bigger things (like this).
The 2nd case is where i always like to get feedback / thoughts from the author - maybe there's a reason/thought behind i didn't grasp so far - and am now introducing another problem i didn't see yet.

Comment (samgermain) [reply tests/test_freqtradebot.py:3551]: ok, sounds good

**Review by xmatthias — APPROVED** [review]
LGTM :+1: 

Hope you're ok with the changes i made ...

**Pull request merged**
